File: Objective-C/04-ViewsAndViewHierarchy/Hypnosister/Hypnosister/HypnosisView.m
//
//  HypnosisView.m
//  Hypnosister
//
//  Created by luckytantanfu on 8/9/15.
//  Copyright (c) 2015 futantan. All rights reserved.
//

#import "HypnosisView.h"

@implementation HypnosisView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        self.backgroundColor = [UIColor clearColor];
    }

    return self;
}


- (void)drawRect:(CGRect)rect {
    CGRect bounds = self.bounds;

    CGPoint center;
    center.x = (CGFloat) (bounds.origin.x + bounds.size.width / 2.0);
    center.y = (CGFloat) (bounds.origin.y + bounds.size.height / 2.0);

    float maxRadius = (float) ((MIN(bounds.size.width, bounds.size.height)) / 2.0);

    UIBezierPath *path = [[UIBezierPath alloc] init];

    for (float currentRadius = maxRadius; currentRadius > 0; currentRadius -= 20) {
        [path moveToPoint:CGPointMake(center.x + currentRadius, center.y)];
        [path addArcWithCenter:center
                        radius:currentRadius
                    startAngle:0.0
                      endAngle:(CGFloat) (M_PI * 2.0)
                     clockwise:YES];
    }

    path.lineWidth = 10;
    [[UIColor lightGrayColor] setStroke];

    [path stroke];

    float imageHeight = 200;
    float imageWidth = 150;
    CGRect imageRect = CGRectMake(
            (CGFloat) ((bounds.size.width - imageWidth) / 2.0),
            (CGFloat) ((bounds.size.height - imageHeight) / 2.0),
            imageWidth,
            imageHeight);

    UIImage *logoImage = [UIImage imageNamed:@"logo.png"];
    [logoImage drawInRect:imageRect];
}

@end

File: Swift/04-ViewsAndViewHierarchy/Hypnosister/Hypnosister/HypnosisView.swift
//
//  HypnosisView.swift
//  Hypnosister
//
//  Created by luckytantanfu on 8/10/15.
//  Copyright (c) 2015 futantan. All rights reserved.
//

import UIKit

class HypnosisView: UIView {
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.backgroundColor = UIColor.clearColor()
    }
    
    required init(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }

    override func drawRect(rect: CGRect) {
        let bounds = self.bounds
        let center = CGPoint(x: bounds.origin.x + bounds.size.width / 2.0, y: bounds.origin.y + bounds.size.height / 2.0)
        let maxRadius = min(bounds.size.width, bounds.size.height) / 2.0
        
        let path = UIBezierPath()
        
        for (var currentRadius = maxRadius; currentRadius > 0; currentRadius = currentRadius - CGFloat(20)) {
            path.moveToPoint(CGPointMake(bounds.width/2.0 + currentRadius, bounds.height/2.0))
            path.addArcWithCenter(center, radius: currentRadius, startAngle: CGFloat(0.0), endAngle: CGFloat(M_PI * 2.0), clockwise: true)
        }
        
        path.lineWidth = 10
        UIColor.lightGrayColor().setStroke()
        path.stroke()
        
        let logoImage = UIImage(named: "logo")
        logoImage?.drawInRect(bounds)
    }
}
